examples/common/convert_c4.py

```diff
@@ -4,105 +4,344 @@
+
+        # note, you need to copy the numpy array because the original is non-writeable
+        # and torch does not support non-writeable tensors, so you get a scary warning and
+        # if you do try to write to the tensor you get undefined behavior
+        tokens = torch.from_numpy(np.frombuffer(ds[0]['tokens'], dtype=np.int64).copy())
+        print(tokenizer.decode(tokens))
+    ```
+    """
+
+    def __init__(
+        self,
+        split: str,
+        tokenizer: PreTrainedTokenizerBase,
+        max_length: int,
+        bos_text: str,
+        eos_text: str,
+        no_wrap: bool,
+    ):
+        self.tokenizer = tokenizer
+        os.environ['TOKENIZERS_PARALLELISM'] = 'false'
+        self.max_length = max_length
+        self.bos_text = bos_text
+        self.eos_text = eos_text
+        self.should_wrap = not no_wrap
+        self.hf_dataset = hf_datasets.load_dataset(path='c4',
+                                                   name='en',
+                                                   split=split,
+                                                   streaming=True)
+
+        self.bos_tokens = self.tokenizer(self.bos_text,
+                                         truncation=False,
+                                         padding=False,
+                                         add_special_tokens=False)['input_ids']
+        if len(self.bos_tokens) > 1:
+            warnings.warn(
+                f'You specified --concat_tokens with --bos_text, but your BOS text is not tokenizing to one token\
+                , instead we got {self.bos_tokens}. Quit if this was in error.')
+
+        self.eos_tokens = self.tokenizer(self.eos_text,
+                                         truncation=False,
+                                         padding=False,
+                                         add_special_tokens=False)['input_ids']
+        if len(self.eos_tokens) > 1:
+            warnings.warn(
+                f'You specified --concat_tokens with --eos_text, but your EOS text is not tokenizing to one token\
+                , instead we got {self.eos_tokens}. Quit if this was in error.')
+
+    def __iter__(self) -> Iterable[Dict[str, bytes]]:
+
+        buffer = []
+        for sample in self.hf_dataset:
+            encoded = self.tokenizer(sample['text'],
+                                     truncation=False,
+                                     padding=False)
+            iids = encoded['input_ids']
+            buffer = buffer + self.bos_tokens + iids + self.eos_tokens
+            while len(buffer) >= self.max_length:
+                concat_sample = buffer[:self.max_length]
+                buffer = buffer[self.max_length:] if self.should_wrap else []
+                yield {
+                    # convert to bytes to store in MDS binary format
+                    'tokens': np.asarray(concat_sample).tobytes()
+                }
+
+
+def build_hf_c4_dataset(
```

Comment:
most of `_get_kwargs` output is used by `build_hf_c4_dataset`.
`build_hf_c4_dataset` should absorb `_get_kwargs`.

Reply:
agreed, will do

Reply:
ok refactored. still testing it.